tsfc/impero.py

```diff
@@ -24,10 +26,22 @@ class Node(NodeBase):
 class Terminal(Node):
     """Abstract class for terminal Impero nodes"""
 
+    __metaclass__ = ABCMeta
+
     __slots__ = ()
 
     children = ()
 
+    @abstractmethod
+    def loop_shape(self, free_indices):
+        """Gives the loop shape, an ordering of indices for an Impero
+        terminal.
+
+        :arg free_indices: a mapping of GEM expressions to ordered
+                           free indices.
```

Comment:
It looks like the usage suggests that `free_indices` is used as a function that takes an expression and returns some ordered free indices for that expression.